src/main/java/org/radarcns/monitor/DisconnectMonitor.java

```diff
@@ -77,13 +124,24 @@ protected void evaluateRecords(ConsumerRecords<GenericRecord, GenericRecord> rec
         Iterator<Map.Entry<String, Long>> iterator = state.lastSeen.entrySet().iterator();
         while (iterator.hasNext()) {
             Map.Entry<String, Long> entry = iterator.next();
-            long timeout = now - entry.getValue();
+            // calculate timeout from current timestamp per device
+            long lastSeen = entry.getValue();
+            long timeout = now - lastSeen;
             if (timeout > timeUntilReportedMissing) {
                 String missingKey = entry.getKey();
+                // remove processed records to prevent repeatitive alerts
                 iterator.remove();
                 reportMissing(stringToKey(missingKey), timeout);
-                state.reportedMissing.put(missingKey, now);
+                try {
+                    // store last seen and reportedMissing timestamp
+                    state.reportedMissing.put(missingKey,
+                            missingRecordsReportToString(new MissingRecordsReport(lastSeen, now)));
```

Comment:
It should only be converted to `String` if it is a key. For values, no such conversion is necessary.